config/config.go

```diff
@@ -106,69 +106,86 @@ func parsePlugins(pluginList []interface{}) ([]*PluginConfig, error) {
 			args = strings.Fields(cast.ToString(v))
 			break
 		}
-		plugins = append(plugins, &PluginConfig{Name: name, Args: args})
+		plugins = append(plugins, PluginConfig{Name: name, Args: args})
 	}
 	return plugins, nil
 }
 
-func (c *Config) getInterface(ver protocolVersion) (string, error) {
-	if err := protoVersionCheck(ver); err != nil {
-		return "", err
-	}
-	directive := fmt.Sprintf("server%d.interface", ver)
-	if exists := c.v.Get(directive); exists == nil {
-		return "", ConfigErrorFromString("dhcpv%d: missing `%s` directive", ver, directive)
+// BUG(Natolumin): listen specifications of the form `[ip6]%iface:port` or
+// `[ip6]%iface` are not supported, even though they are the default format of
+// the `ss` utility in linux. Use `[ip6%iface]:port` instead
+
+// splitHostPort splits an address of the form ip%zone:port into ip,zone and port.
+// It still returns if any of these are unset (unlike net.SplitHostPort which
+// returns an error if there is no port)
+func splitHostPort(hostport string) (ip string, zone string, port string, err error) {
+	ip, port, err = net.SplitHostPort(hostport)
+	if err != nil {
+		// Either there is no port, or a more serious error.
+		// Supply a synthetic port to differentiate cases
+		var altErr error
+		if ip, _, altErr = net.SplitHostPort(hostport + ":0"); altErr != nil {
+			// Invalid even with a fake port. Return the original error
+			return
+		}
+		err = nil
 	}
-	ifname := c.v.GetString(directive)
-	if ifname == "" {
-		return "", ConfigErrorFromString("dhcpv%d: missing `%s` directive", ver, directive)
+	if i := strings.LastIndexByte(ip, '%'); i >= 0 {
+		ip, zone = ip[:i], ip[i+1:]
 	}
-	return ifname, nil
+	return
 }
 
-func (c *Config) getListenAddress(ver protocolVersion) (*net.UDPAddr, error) {
+func (c *Config) getListenAddress(addr string, ver protocolVersion) (*net.UDPAddr, error) {
 	if err := protoVersionCheck(ver); err != nil {
 		return nil, err
 	}
-	addr := c.v.GetString(fmt.Sprintf("server%d.listen", ver))
-	if addr == "" {
-		// return default listener
-		if ver == protocolV6 {
-			return &net.UDPAddr{
-				IP:   net.IPv6unspecified,
-				Port: dhcpv6.DefaultServerPort,
-			}, nil
-		}
-		return &net.UDPAddr{
-			IP:   net.IPv4zero,
-			Port: dhcpv4.ServerPort,
-		}, nil
-	}
-	ipStr, portStr, err := net.SplitHostPort(addr)
+
+	ipStr, ifname, portStr, err := splitHostPort(addr)
 	if err != nil {
 		return nil, ConfigErrorFromString("dhcpv%d: %v", ver, err)
 	}
+
 	ip := net.ParseIP(ipStr)
+	if ipStr == "" {
+		switch ver {
+		case protocolV4:
+			ip = net.IPv4zero
+		case protocolV6:
+			ip = net.IPv6unspecified
+		}
```

Comment:
Should we have a default here?

Reply:
Since we're inside config/ and the `ver` variable we're switching over never comes from user input, taking the default would always indicate a bug. Even then, since `ipStr == ""`, `ParseIP` will have returned nil and we enter the `if ip == nil` right after this and error out anyway.

So not having a default is not an error; however we can decide it's better for clarity and add one. I don't think it's necessary, but I'm not opposed to it so it's your call

Reply:
yeah, it would indicate a bug. I am OK with either calling a `panic()` here, and be loud about the bug, or silently going ahead given how unlikely this case is